File: src/controls/control.ts
import type { Responder } from "../core/responder";
import { View } from "../core/view";

export interface ControlDelegate {
  controlShouldResignFirstResponder?(control: Control, next: Responder | null): boolean;
}

export abstract class Control extends View {
  delegate?: ControlDelegate;

  override resignFirstResponder(next: Responder | null = null): boolean {
    if (this.delegate?.controlShouldResignFirstResponder?.(this, next) === false) {
      return false;
    }
    return super.resignFirstResponder(next);
  }
}

File: opal/lib/swill/control.rb
# frozen_string_literal: true

module Swill
  # Base form control view. Delegates may veto first-responder resignation.
  class Control < View
    attr_accessor :delegate

    def resign_first_responder(next_responder = nil)
      if delegate&.respond_to?(:control_should_resign_first_responder) &&
         delegate.control_should_resign_first_responder(self, next_responder) == false
        return false
      end

      super
    end
  end
end
